app/javascript/controllers/file_input_controller.js

```diff
@@ -0,0 +1,48 @@
+import { Controller } from "@hotwired/stimulus";
+
+/**
+ * Stimulus controller to enhance the file input UI for multiple files.
+ *
+ * This controller:
+ * - Listens for file selection on an `<input type="file" multiple="multiple">`
+ * - Displays selected file names in a custom list **when multiple files are selected
+ * - Defaults to the browser’s built-in display for a single file selection
+ *
+ * Expected HTML structure should have a placeholder div for the selected file names:
+ *
+ * ```erb
+ * <div data-controller="file-input">
+ *   <input type="file" multiple data-file-input-target="input">
+ *   <div data-file-input-target="list"></div>
+ * </div>
+ * ```
+ */
+export default class extends Controller {
+  static targets = ["input", "list"];
+
+  connect() {
+    this.inputTarget.addEventListener("change", () => this.updateFileList());
+  }
+
+  updateFileList() {
+    const files = this.inputTarget.files;
+    this.listTarget.innerHTML = ""; // Clear previous list
+
+    // If no files or only one file is selected, let the native UI handle it
+    if (files.length <= 1) {
+      return;
+    }
+
+    const ul = document.createElement("ul");
+    ul.classList.add("list-unstyled", "mt-2");
```

Comment:
This is a fine way to do it, but thinking to myself here I wish we didn't need the `mt-2`; seems like it might later be harder to maintain.

Reply:
I tried removing the `mt-2` and it looks fine. This has already been merged but there's still a Part 3 for this issue, so I can remove the extra margin in that PR.